File: productions/planet-potion/work/re/texbisect.py
#!/usr/bin/env python3
"""Find the FIRST opcode in a program whose float surface diverges from the JS.

    python3 texbisect.py flat/ dataset/ p1_15

A whole-program diff says "this texture is wrong by three subpixels" and leaves
the search to guesswork over a dozen composed opcodes. This runs every prefix of
the program — one opcode, then two, then three — through the original and writes
each surface as float32, so the JS side can compare prefix by prefix and name the
first one that differs. Floats rather than bytes, because a difference that has
not yet crossed a rounding boundary is invisible in the texture and still wrong.

Writes <out>/<name>_<k>.json for each prefix; texbisect.mjs reads them back.
"""
import json
import os
import struct
import sys

import ppcrun as H
import texfloat
import texopref
from texopsuite import decode


def prefix_blob(ops, k):
    body = b''.join(bytes([op]) + bytes(operands) for op, operands in ops[:k])
    return struct.pack('>H', len(body)) + body


def main():
    flat, dataset, name = sys.argv[1], sys.argv[2], sys.argv[3]
    out = sys.argv[4] if len(sys.argv) > 4 else 'texbisect'
    os.makedirs(out, exist_ok=True)
    progs = {f"{p['part']}_{p['index']}": p
             for p in json.load(open(f'{dataset}/tex_programs.json'))['programs']}
    ops = decode(bytes.fromhex(progs[name]['hex']))

    d0 = open(f'{flat}/seg0_CODE_10000000.bin', 'rb').read()
    H.preload_tables(d0)
    addr = texfloat.surface_ptr(d0, 'current')
    index = []
    for k in range(1, len(ops) + 1):
        H.PRELOAD[texopref.PROG_AT] = prefix_blob(ops, k)
        raw, err = H.run(flat, H.sym(flat)['_generate'],
                         {3: texopref.PROG_AT, 4: H.DEST},
                         out_addr=addr, out_len=128 * 128 * 16)
        if len(raw) != 128 * 128 * 16:
            raise SystemExit(f'texbisect: prefix {k}: {err[:200]}')
        vals = struct.unpack('>65536f', raw)
        json.dump({'floats': list(vals)}, open(f'{out}/{name}_{k}.json', 'w'))
        op, operands = ops[k - 1]
        index.append({'k': k, 'op': op, 'operands': operands,
                      'file': f'{name}_{k}.json'})
        print(f'{k:2}  op{op if op < 20 else hex(op)}  {len(operands)} operands')

    json.dump({'program': name,
               'ops': [{'op': o, 'operands': v} for o, v in ops],
               'prefixes': index}, open(f'{out}/{name}.json', 'w'), indent=2)
    print(f'\nwrote {len(index)} prefixes to {out}/')


if __name__ == '__main__':
    main()

File: productions/planet-potion/work/re/texbisect.mjs
// texbisect.mjs — name the first opcode whose float surface diverges.
//
//   node work/re/texbisect.mjs /tmp/texbisect p1_15 [--kernels tex_kernels.json]
//
// Reads the prefix surfaces texbisect.py wrote and runs the same prefixes in JS,
// comparing float32 values rather than the 8-bit texture — so a divergence is
// reported at the opcode that introduced it, not at the one that finally pushed
// it across a rounding boundary.
import fs from 'node:fs';
import { run, PIXELS } from '../../web/js/texturevm.js';

const dir = process.argv[2], name = process.argv[3];
const ki = process.argv.indexOf('--kernels');
const kernels = {};
if (ki > 0) {
  const doc = JSON.parse(fs.readFileSync(process.argv[ki + 1], 'utf8'));
  for (const [k, v] of Object.entries(doc.kernels ?? doc)) {
    kernels[Number(k)] = Array.isArray(v) ? v : (v.kernel ?? v.weights);
  }
}

const meta = JSON.parse(fs.readFileSync(`${dir}/${name}.json`, 'utf8'));
for (const p of meta.prefixes) {
  const want = JSON.parse(fs.readFileSync(`${dir}/${p.file}`, 'utf8')).floats;
  const got = run(meta.ops.slice(0, p.k), kernels).current;
  let n = 0, first = null, high = 0;
  for (let i = 0; i < PIXELS * 4; i++) {
    if (got[i] !== want[i]) {
      n++;
      if (got[i] > want[i]) high++;
      if (!first) {
        first = `px ${(i >> 2) % 128},${(i >> 2) / 128 | 0} ch${i & 3}`
          + `  want ${want[i]}  got ${got[i]}`;
      }
    }
  }
  const label = p.op < 20 ? `op${p.op}` : `0x${p.op.toString(16)}`;
  console.log(`${p.k === 1 ? '' : ''}${String(p.k).padStart(2)}  ${label.padEnd(6)}`
    + `${n === 0 ? 'identical' : `${n} floats differ (${high} high)`}`);
  if (n) { console.log(`      ${first}`); break; }
}
